Optional top level decorator functionality

hackgnar opened the pull request:
Added optional functionality to allow decorator data to populate as top level values instead of being contained as a child of decorations.  This can be enabled by setting the optional configuration flag `top_level_decorations` to `true`.  This is useful for setups where data needs tagged or munged before being sent directly from osquery to a log collection destination.

osqueryer commented:
Can one of the osquery admins reply with "ok to test" to kick off a build...

PoppySeedPlehzr commented:
ok to test

osqueryer commented:
Build origin/pr/2177/merge (Job: 2771) is good! :+1:

theopolis commented on osquery/database/database.cpp:
```diff
@@ -35,6 +35,10 @@ CLI_FLAG(bool,
 
 FLAG(bool, disable_database, false, "Disable the persistent RocksDB storage");
+FLAG(bool,
+     top_level_decorators,
```
Can you define this in: `./config/parsers/decorators.cpp` where `disable_decorators` lives?

Then here use:
```cpp
DECLARE_bool(top_level_decorators);
```

theopolis commented on docs/wiki/deployment/configuration.md:
```diff
@@ -429,6 +429,8 @@ The columns, and their values, will be appended to each log line as follows. Ass
 
 Expect the normal set of log keys to be included and note that `decorations` is a top-level key in the log line whose value is an embedded map.
 
+The configuration flag `top_level_decorations`  can be set to `true` to make decorator data populate as top level key/value objects instead of being contained as a child of `decorations`.  When using this feature, you must be weary of key collisions in your data.
```
Can you expand on "collisions in your data"? Maybe "collisions in existing, reserved, top-level keys"?

It would be awesome to provide an example of what this does, perhaps a _with_ and _without_ example log line.

theopolis commented on osquery/database/database.cpp:
```diff
@@ -266,10 +270,16 @@ inline void addLegacyFieldsAndDecorations(const QueryLogItem& item,
-    for (const auto& name : item.decorations) {
-      decorations.put<std::string>(name.first, name.second);
+    if (FLAGS_top_level_decorators) {
+      for (const auto& name : item.decorations) {
```
Can this be collapsed into 1 loop?
```cpp
auto& decorator_parent = tree;
if (!...) {
  tree.add_child("decorations", pt::ptree());
  decorator_parent = tree.get_child("decorations");
}
for (const auto& name : item.decorations) {
  decorator_parent.put<std::string>(name.first, name.second);
}
```

hackgnar replied:
@theopolis I initially tried to do this, but went with an if/else with 2 loops because I couldn't figure out a way to do this without running into reference reassignment issues.  

So in the example above, you initially set the auto reference `decorator_parent` to tree, then when the the top level decorator flag is false, you set `decorator_parent` to a child tree.  The reassignment to the child tree ends up not working when I run this code.

Maybe my brain is not working today.  Am I looking at this wrong?  get_child returns a ptree ref and we cant reassign refs correct?

theopolis commented:
We have some weird conventions, one of them is to prefix the feature category for the CLI flags so `top_level_decorations` would become `decorations_top_level` or `decorations_toplevel`.

Also, do you mind adding a test here: https://github.com/facebook/osquery/blob/master/osquery/config/parsers/tests/decorators_tests.cpp

You can enable the feature when the test function starts and restore the value (or just turn it off) at the end.

hackgnar commented:
Thanks for the advice @theopolis !  Ill add all of your requests ASAP.

osqueryer commented:
Build origin/pr/2177/merge (Job: 2832) is good! :+1:

hackgnar commented:
@theopolis I updated the PR with the following based on your requests:

1. Changed the declaration location of the flag
2. Changed the name of the flag to match conventions
3. Updated documentation of the flag
4. Added a unit test

I couldn't come up with a good solution of the double loop suggestion though.  See my comment on your recommendation for more details.

Let me know if there is anything else you would like me to add.

theopolis commented:
You're right about the re-binding @hackgnar! Though we can emulate the recommendation with `std::ref` (which allows the rebind in C++11).

hackgnar commented:
@theopolis awesome, thx for the tip!  Ill check out `std::ref` first thing in the morning and add the change

osqueryer commented:
Build origin/pr/2177/merge (Job: 2836) is good! :+1:

hackgnar commented:
@theopolis axed the double loop with std::ref.  Cool C++11 feature, thx for the advice.  Let me know if everything is gtg.

zwass commented:
Thanks @hackgnar.
